File: aeroport.py
"""
Module Aeroport - Gère l'ensemble du système de contrôle aérien

LOGIQUE:
L'aéroport coordonne:
1. La gestion des pistes (disponibilité, attribution)
2. La file d'attente des avions en approche
3. Les avions au sol prêts à décoller
4. L'attribution intelligente des pistes aux avions

Algorithme d'attribution:
- Priorité aux atterrissages (sécurité)
- Attribution de la première piste disponible
- Gestion de la file d'attente si toutes les pistes sont occupées
"""

from avion import Avion, StatutAvion
from piste import Piste


class Aeroport:
    """
    Classe représentant un aéroport avec son système de contrôle aérien
    
    Attributes:
        nom (str): Nom de l'aéroport
        pistes (list): Liste des pistes disponibles
        avions_en_approche (list): File d'attente des avions en approche
        avions_au_sol (list): Liste des avions au sol
        historique (list): Historique des événements
    """
    
    def __init__(self, nom, nombre_pistes=3):
        """
        Initialise un nouvel aéroport
        
        Args:
            nom (str): Nom de l'aéroport
            nombre_pistes (int): Nombre de pistes à créer (défaut: 3)
        """
        self.nom = nom
        self.pistes = [Piste(i + 1) for i in range(nombre_pistes)]
        self.avions_en_approche = []
        self.avions_au_sol = []
        self.historique = []
    
    def ajouter_avion_en_approche(self, avion):
        """
        Ajoute un avion à la file d'attente des approches
        
        Args:
            avion (Avion): L'avion en approche
        """
        avion.changer_statut(StatutAvion.EN_APPROCHE)
        self.avions_en_approche.append(avion)
        message = f"✈️  Vol {avion.code_vol} signale son approche de l'aéroport"
        self.historique.append(message)
        print(message)
    
    def trouver_piste_disponible(self):
        """
        Trouve la première piste disponible
        
        Returns:
            Piste: Une piste disponible, ou None si aucune n'est disponible
        """
        for piste in self.pistes:
            if piste.est_disponible():
                return piste
        return None
    
    def autoriser_atterrissage(self):
        """
        Autorise l'atterrissage du prochain avion en approche
        
        LOGIQUE:
        1. Vérifie s'il y a des avions en attente
        2. Cherche une piste disponible
        3. Assigne la piste et change le statut de l'avion
        4. Simule l'atterrissage
        
        Returns:
            bool: True si un atterrissage a été autorisé, False sinon
        """
        if not self.avions_en_approche:
            print("⚠️  Aucun avion en approche")
            return False
        
        piste = self.trouver_piste_disponible()
        if not piste:
            print("⚠️  Toutes les pistes sont occupées, l'avion doit patienter")
            return False
        
        # Traiter le premier avion en attente
        avion = self.avions_en_approche.pop(0)
        avion.changer_statut(StatutAvion.ATTERRISSAGE)
        avion.assigner_piste(piste.numero)
        piste.occuper(avion)
        
        message = f"🛬 Vol {avion.code_vol} autorisé à atterrir sur piste {piste.numero}"
        self.historique.append(message)
        print(message)
        
        return True
    
    def finaliser_atterrissage(self, code_vol):
        """
        Finalise l'atterrissage d'un avion (libère la piste, avion au sol)
        
        Args:
            code_vol (str): Code du vol à finaliser
            
        Returns:
            bool: True si la finalisation a réussi, False sinon
        """
        # Chercher l'avion sur une piste
        for piste in self.pistes:
            if (piste.avion_actuel and 
                piste.avion_actuel.code_vol == code_vol and
                piste.avion_actuel.statut == StatutAvion.ATTERRISSAGE):
                
                avion = piste.liberer()
                avion.changer_statut(StatutAvion.AU_SOL)
                avion.liberer_piste()
                self.avions_au_sol.append(avion)
                
                message = f"✅ Vol {code_vol} a atterri et libéré la piste {piste.numero}"
                self.historique.append(message)
                print(message)
                return True
        
        print(f"⚠️  Vol {code_vol} non trouvé en phase d'atterrissage")
        return False
    
    def autoriser_decollage(self, code_vol):
        """
        Autorise le décollage d'un avion au sol
        
        Args:
            code_vol (str): Code du vol à faire décoller
            
        Returns:
            bool: True si le décollage a été autorisé, False sinon
        """
        # Chercher l'avion au sol
        avion = None
        for a in self.avions_au_sol:
            if a.code_vol == code_vol:
                avion = a
                break
        
        if not avion:
            print(f"⚠️  Vol {code_vol} non trouvé au sol")
            return False
        
        # Chercher une piste disponible
        piste = self.trouver_piste_disponible()
        if not piste:
            print("⚠️  Toutes les pistes sont occupées, impossible de décoller maintenant")
            return False
        
        # Autoriser le décollage
        self.avions_au_sol.remove(avion)
        avion.changer_statut(StatutAvion.DECOLLAGE)
        avion.assigner_piste(piste.numero)
        piste.occuper(avion)
        
        message = f"🛫 Vol {code_vol} autorisé à décoller de la piste {piste.numero}"
        self.historique.append(message)
        print(message)
        
        return True
    
    def finaliser_decollage(self, code_vol):
        """
        Finalise le décollage d'un avion (libère la piste, avion en vol)
        
        Args:
            code_vol (str): Code du vol à finaliser
            
        Returns:
            bool: True si la finalisation a réussi, False sinon
        """
        # Chercher l'avion sur une piste
        for piste in self.pistes:
            if (piste.avion_actuel and 
                piste.avion_actuel.code_vol == code_vol and
                piste.avion_actuel.statut == StatutAvion.DECOLLAGE):
                
                avion = piste.liberer()
                avion.changer_statut(StatutAvion.EN_VOL)
                avion.liberer_piste()
                
                message = f"✅ Vol {code_vol} a décollé et libéré la piste {piste.numero}"
                self.historique.append(message)
                print(message)
                return True
        
        print(f"⚠️  Vol {code_vol} non trouvé en phase de décollage")
        return False
    
    def afficher_etat(self):
        """Affiche l'état complet de l'aéroport"""
        print("\n" + "="*60)
        print(f"📍 AÉROPORT: {self.nom}")
        print("="*60)
        
        print("\n🛫 PISTES:")
        for piste in self.pistes:
            print(f"  {piste}")
        
        print(f"\n✈️  AVIONS EN APPROCHE ({len(self.avions_en_approche)}):")
        if self.avions_en_approche:
            for avion in self.avions_en_approche:
                print(f"  - {avion}")
        else:
            print("  Aucun")
        
        print(f"\n🅿️  AVIONS AU SOL ({len(self.avions_au_sol)}):")
        if self.avions_au_sol:
            for avion in self.avions_au_sol:
                print(f"  - {avion}")
        else:
            print("  Aucun")
        
        print("="*60 + "\n")
    
    def afficher_historique(self):
        """Affiche l'historique des événements"""
        print("\n" + "="*60)
        print("📋 HISTORIQUE DES ÉVÉNEMENTS")
        print("="*60)
        for i, event in enumerate(self.historique, 1):
            print(f"{i}. {event}")
        print("="*60 + "\n")

File: java/src/Aeroport.java
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;

/**
 * Aeroport - Cœur du système de contrôle aérien.
 *
 * LOGIQUE GÉNÉRALE:
 * L'aéroport coordonne trois ressources :
 *   1. {@code pistes}           – liste des pistes physiques (nombre fixé à la création)
 *   2. {@code avionsEnApproche} – file d'attente FIFO des avions qui veulent atterrir
 *   3. {@code avionsAuSol}      – liste des avions garés, prêts à décoller
 *
 * ALGORITHME D'ATTRIBUTION DE PISTE:
 *   - On parcourt les pistes dans l'ordre et on prend la première disponible.
 *   - Si toutes les pistes sont occupées, l'opération est refusée (retour false).
 *
 * PRIORITÉS:
 *   - Les atterrissages sont traités en FIFO (premier arrivé = premier servi).
 *   - Les décollages sont initiés à la demande explicite (code vol fourni).
 *   - Atterrissage prioritaire sur décollage en cas de conflit de ressource.
 *
 * FLUX D'ATTERRISSAGE:
 *   ajouter → autoriserAtterrissage → finaliserAtterrissage
 *
 * FLUX DE DÉCOLLAGE:
 *   autoriserDecollage(codeVol) → finaliserDecollage(codeVol)
 */
public class Aeroport {

    private final String nom;
    private final List<Piste> pistes;
    private final Queue<Avion> avionsEnApproche;  // FIFO
    private final List<Avion> avionsAuSol;
    private final List<String> historique;

    /**
     * Crée un aéroport avec le nombre de pistes indiqué.
     *
     * @param nom          nom de l'aéroport
     * @param nombrePistes nombre de pistes à créer (doit être >= 1)
     */
    public Aeroport(String nom, int nombrePistes) {
        this.nom = nom;
        this.pistes = new ArrayList<>();
        for (int i = 1; i <= nombrePistes; i++) {
            pistes.add(new Piste(i));
        }
        this.avionsEnApproche = new LinkedList<>();
        this.avionsAuSol = new ArrayList<>();
        this.historique = new ArrayList<>();
    }

    /** Constructeur avec 3 pistes par défaut. */
    public Aeroport(String nom) {
        this(nom, 3);
    }

    // ------------------------------------------------------------------ //
    //  Getters (utiles pour les tests)
    // ------------------------------------------------------------------ //

    public String getNom() { return nom; }
    public List<Piste> getPistes() { return pistes; }
    public Queue<Avion> getAvionsEnApproche() { return avionsEnApproche; }
    public List<Avion> getAvionsAuSol() { return avionsAuSol; }
    public List<String> getHistorique() { return historique; }

    // ------------------------------------------------------------------ //
    //  Logique métier
    // ------------------------------------------------------------------ //

    /**
     * Enregistre un avion en approche et l'ajoute à la file d'attente.
     *
     * @param avion l'avion en approche
     */
    public void ajouterAvionEnApproche(Avion avion) {
        avion.changerStatut(StatutAvion.EN_APPROCHE);
        avionsEnApproche.add(avion);
        String msg = "✈  Vol " + avion.getCodeVol() + " signale son approche de l'aéroport";
        historique.add(msg);
        System.out.println(msg);
    }

    /**
     * Cherche la première piste disponible parmi toutes les pistes.
     *
     * @return une piste disponible, ou null si toutes sont occupées/en maintenance
     */
    public Piste trouverPisteDisponible() {
        for (Piste piste : pistes) {
            if (piste.estDisponible()) {
                return piste;
            }
        }
        return null;
    }

    /**
     * Autorise l'atterrissage du prochain avion en file d'attente.
     *
     * LOGIQUE:
     *  1. S'il n'y a aucun avion en approche → refus.
     *  2. Si aucune piste n'est libre       → refus (avion reste en tête de file).
     *  3. Sinon : défile le premier avion, lui assigne une piste,
     *     change son statut en ATTERRISSAGE et occupe la piste.
     *
     * @return true si un atterrissage a été autorisé, false sinon
     */
    public boolean autoriserAtterrissage() {
        if (avionsEnApproche.isEmpty()) {
            System.out.println("⚠  Aucun avion en approche");
            return false;
        }

        Piste piste = trouverPisteDisponible();
        if (piste == null) {
            System.out.println("⚠  Toutes les pistes sont occupées, l'avion doit patienter");
            return false;
        }

        // Défile le premier avion (FIFO)
        Avion avion = avionsEnApproche.poll();
        avion.changerStatut(StatutAvion.ATTERRISSAGE);
        avion.assignerPiste(piste.getNumero());
        piste.occuper(avion);

        String msg = ">> Vol " + avion.getCodeVol() + " autorisé à atterrir sur piste " + piste.getNumero();
        historique.add(msg);
        System.out.println(msg);
        return true;
    }

    /**
     * Finalise l'atterrissage d'un avion : libère la piste, pose l'avion au sol.
     *
     * @param codeVol code du vol dont l'atterrissage est terminé
     * @return true si la finalisation a réussi, false si le vol n'est pas trouvé
     */
    public boolean finaliserAtterrissage(String codeVol) {
        for (Piste piste : pistes) {
            Avion avion = piste.getAvionActuel();
            if (avion != null
                    && avion.getCodeVol().equals(codeVol)
                    && avion.getStatut() == StatutAvion.ATTERRISSAGE) {

                piste.liberer();
                avion.changerStatut(StatutAvion.AU_SOL);
                avion.libererPiste();
                avionsAuSol.add(avion);

                String msg = "OK Vol " + codeVol + " a atterri et libéré la piste " + piste.getNumero();
                historique.add(msg);
                System.out.println(msg);
                return true;
            }
        }
        System.out.println("⚠  Vol " + codeVol + " non trouvé en phase d'atterrissage");
        return false;
    }

    /**
     * Autorise le décollage d'un avion identifié par son code de vol.
     *
     * LOGIQUE:
     *  1. Cherche l'avion dans la liste des avions au sol.
     *  2. Cherche une piste disponible.
     *  3. Si les deux conditions sont remplies, change le statut et occupe la piste.
     *
     * @param codeVol code du vol à faire décoller
     * @return true si le décollage a été autorisé, false sinon
     */
    public boolean autoriserDecollage(String codeVol) {
        Avion avion = trouverAvionAuSol(codeVol);
        if (avion == null) {
            System.out.println("⚠  Vol " + codeVol + " non trouvé au sol");
            return false;
        }

        Piste piste = trouverPisteDisponible();
        if (piste == null) {
            System.out.println("⚠  Toutes les pistes sont occupées, impossible de décoller maintenant");
            return false;
        }

        avionsAuSol.remove(avion);
        avion.changerStatut(StatutAvion.DECOLLAGE);
        avion.assignerPiste(piste.getNumero());
        piste.occuper(avion);

        String msg = ">> Vol " + codeVol + " autorisé à décoller de la piste " + piste.getNumero();
        historique.add(msg);
        System.out.println(msg);
        return true;
    }

    /**
     * Finalise le décollage d'un avion : libère la piste, l'avion quitte l'espace aérien.
     *
     * @param codeVol code du vol dont le décollage est terminé
     * @return true si la finalisation a réussi, false sinon
     */
    public boolean finaliserDecollage(String codeVol) {
        for (Piste piste : pistes) {
            Avion avion = piste.getAvionActuel();
            if (avion != null
                    && avion.getCodeVol().equals(codeVol)
                    && avion.getStatut() == StatutAvion.DECOLLAGE) {

                piste.liberer();
                avion.changerStatut(StatutAvion.EN_VOL);
                avion.libererPiste();

                String msg = "OK Vol " + codeVol + " a décollé et libéré la piste " + piste.getNumero();
                historique.add(msg);
                System.out.println(msg);
                return true;
            }
        }
        System.out.println("⚠  Vol " + codeVol + " non trouvé en phase de décollage");
        return false;
    }

    // ------------------------------------------------------------------ //
    //  Affichage
    // ------------------------------------------------------------------ //

    /** Affiche l'état complet de l'aéroport dans la console. */
    public void afficherEtat() {
        String sep = "=".repeat(60);
        System.out.println();
        System.out.println(sep);
        System.out.println("AEROPORT: " + nom);
        System.out.println(sep);

        System.out.println("\nPISTES:");
        for (Piste p : pistes) {
            System.out.println("  " + p);
        }

        System.out.println("\nAVIONS EN APPROCHE (" + avionsEnApproche.size() + "):");
        if (avionsEnApproche.isEmpty()) {
            System.out.println("  Aucun");
        } else {
            for (Avion a : avionsEnApproche) {
                System.out.println("  - " + a);
            }
        }

        System.out.println("\nAVIONS AU SOL (" + avionsAuSol.size() + "):");
        if (avionsAuSol.isEmpty()) {
            System.out.println("  Aucun");
        } else {
            for (Avion a : avionsAuSol) {
                System.out.println("  - " + a);
            }
        }

        System.out.println(sep);
        System.out.println();
    }

    /** Affiche le journal chronologique de tous les événements. */
    public void afficherHistorique() {
        String sep = "=".repeat(60);
        System.out.println();
        System.out.println(sep);
        System.out.println("HISTORIQUE DES EVENEMENTS");
        System.out.println(sep);
        int i = 1;
        for (String event : historique) {
            System.out.println(i++ + ". " + event);
        }
        System.out.println(sep);
        System.out.println();
    }

    // ------------------------------------------------------------------ //
    //  Méthodes utilitaires privées
    // ------------------------------------------------------------------ //

    /**
     * Recherche un avion dans la liste des avions au sol.
     *
     * @param codeVol code du vol recherché
     * @return l'avion trouvé, ou null
     */
    private Avion trouverAvionAuSol(String codeVol) {
        for (Avion a : avionsAuSol) {
            if (a.getCodeVol().equals(codeVol)) {
                return a;
            }
        }
        return null;
    }
}
